internal/db/resource.go

```diff
@@ -106,3 +104,68 @@ func (d *Database) ListResources(ctx context.Context, limit, offset int) (_ []Re
 	}
 	return resources, nil
 }
+
+// buildQuery is an utility function to build the database query according to two optional parameters.
+// nameFilter: used to match resources name.
+// typeFilter used to match resources type. If this is not empty the resources are fetched from a single table.
+func buildQuery(db *gorm.DB, offset, limit int, nameFilter, typeFilter string) (*gorm.DB, error) {
+	query := `
+	? UNION ? UNION ? UNION ? UNION ?
+	ORDER BY type, id
+	OFFSET ?
+	LIMIT  ?;
+	`
+	applicationOffersQuery := db.Select(SELECT_APPLICATION_OFFERS).
+		Model(&dbmodel.ApplicationOffer{}).
+		Joins("JOIN models ON application_offers.model_id = models.id")
+
+	cloudsQuery := db.Select(SELECT_CLOUDS).
+		Model(&dbmodel.Cloud{})
+
+	controllersQuery := db.Select(SELECT_CONTROLLERS).
+		Model(&dbmodel.Controller{})
+
+	modelsQuery := db.Select(SELECT_MODELS).
+		Model(&dbmodel.Model{}).
+		Joins("JOIN controllers ON models.controller_id = controllers.id")
+
+	serviceAccountsQuery := db.Select(SELECT_IDENTITIES).
+		Model(&dbmodel.Identity{}).
+		Where("name LIKE '%@serviceaccount'")
+
+	queries := map[string]*gorm.DB{
+		ApplicationOffersQueryKey: applicationOffersQuery,
+		CloudsQueryKey:            cloudsQuery,
+		ControllersQueryKey:       controllersQuery,
+		ModelsQueryKey:            modelsQuery,
+		ServiceAccountQueryKey:    serviceAccountsQuery,
+	}
+	// we add the where clause only if the nameFilter is filled
+	if nameFilter != "" {
+		nameFilter = "%" + nameFilter + "%"
+		for k := range queries {
+			queries[k] = queries[k].Where(k+".name LIKE ?", nameFilter)
```

Comment:
To be honest, I don't like playing with queries like this. I'd rather have the const queries richer with parameters and here just decide on the arguments to pass to them. For example, we can define the const queries like this:

```
SELECT * ...
WHERE ? IS TRUE OR name LIKE ?
```

Where first parameter is a boolean that we set to `false` for prefix matching to take place.

Reply:
using the where in this way simplified the code, thanks for the suggestion